[AMDGPU] Allow casts between the Global and Constant Addr Spaces in isValidAddrSpaceCast

ritter-x2a opened the pull request:
So far, `isValidAddrSpaceCast` only allows casts to the flat address space and between the constant(32) address spaces. It does not allow casting between the global and constant address spaces, even though they alias. That affects, e.g., the lowering of memmoves from the constant to the global address space in LowerMemIntrinsics, since that requires aliasing address spaces to be castable.

This patch allows such casts. It also includes a memmove test that would crash with the previous implementation because the memmove IR lowering would not be applicable for the move from constant AS to global AS.

ritter-x2a committed:
[AMDGPU] Allow casts between the Global and Constant Addr Spaces in isValidAddrSpaceCast

So far, isValidAddrSpaceCast only allows casts to the flat address space and
between the constant(32) address spaces. It does not allow casting between the
global and constant address spaces, even though the constant AS is only an
alias for the global AS. That affects, e.g., the lowering of memmoves from the
constant to the global address space in LowerMemIntrinsics, since that requires
aliasing address spaces to be castable.

This patch allows such casts. It also includes a memmove test that would crash
with the previous implementation because the memmove IR lowering would not be
applicable for the move from constant AS to global AS.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.h:
```diff
@@ -187,9 +187,12 @@ class GCNTTIImpl final : public BasicTTIImplBase<GCNTTIImpl> {
-         ToAS == AMDGPUAS::CONSTANT_ADDRESS) ||
-        (FromAS == AMDGPUAS::CONSTANT_ADDRESS &&
+    if (FromAS != ToAS &&
+        (FromAS == AMDGPUAS::GLOBAL_ADDRESS ||
+         FromAS == AMDGPUAS::CONSTANT_ADDRESS ||
+         FromAS == AMDGPUAS::CONSTANT_ADDRESS_32BIT) &&
+        (ToAS == AMDGPUAS::GLOBAL_ADDRESS ||
+         ToAS == AMDGPUAS::CONSTANT_ADDRESS ||
          ToAS == AMDGPUAS::CONSTANT_ADDRESS_32BIT))
```
The whole function should match isExtendedGlobalAddrSpace. We should accept arbitrary 64-bit address spaces as an alias for global. The part above under ToAS == AMDGPUAS::FLAT_ADDRESS is also too strict.

The only illegal cases involve region and anything, fat pointers and anything, or any of the 32-bit address spaces between each other.

ritter-x2a replied:
Would it make sense to have a table like for [the aliasing rules](https://github.com/llvm/llvm-project/blob/a3b0c31ebc2f6fe672f08f6b7d15f25a2b26edda/llvm/lib/Target/AMDGPU/AMDGPU.h#L473) for this, then?

arsenm replied:
It's a stylistic choice, whatever comes out simpler

ritter-x2a replied:
Adjusted in the most recent commit.

ritter-x2a committed:
fixup! [AMDGPU] Allow casts between the Global and Constant Addr Spaces in isValidAddrSpaceCast

Relax and reorganize isValidAddrSpaceCast further

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.h:
```diff
@@ -174,24 +174,45 @@ class GCNTTIImpl final : public BasicTTIImplBase<GCNTTIImpl> {
+    if (FromAS == ToAS)
+      return false;
+
+    if (FromAS == AMDGPUAS::FLAT_ADDRESS ||
+        FromAS == AMDGPUAS::GLOBAL_ADDRESS ||
+        FromAS == AMDGPUAS::CONSTANT_ADDRESS ||
+        FromAS > AMDGPUAS::MAX_AMDGPU_ADDRESS) {
```
Use AMDGPU::isFlatGlobalAddrSpace instead of repeating it

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.h:
```diff
@@ -174,24 +174,45 @@ class GCNTTIImpl final : public BasicTTIImplBase<GCNTTIImpl> {
+    } else if (FromAS == AMDGPUAS::LOCAL_ADDRESS ||
+               FromAS == AMDGPUAS::PRIVATE_ADDRESS ||
+               FromAS == AMDGPUAS::CONSTANT_ADDRESS_32BIT) {
+      // Casting from a 32-bit AS to any 64-bit AS is valid.
+      switch (ToAS) {
+      case AMDGPUAS::FLAT_ADDRESS:
+      case AMDGPUAS::GLOBAL_ADDRESS:
+      case AMDGPUAS::CONSTANT_ADDRESS:
+        return true;
+      default:
+        break;
```
CONSTANT_ADDRESS_32BIT -> CONSTANT, GLOBAL, or FLAT is ok.

This incorrectly permits local->non-flat and private->non-flat

ritter-x2a replied:
Since addrspace casts are assumed to be reversible, should the inverse directions, i.e., non-flat -> (private or local), then also be forbidden?

ritter-x2a replied:
Assuming that is the case, I updated the PR with 0d29726202dffc47f000520ee648afece9bf3421 accordingly.

arsenm replied:
Yes, p1 -> p3/p5 are also illegal

ritter-x2a committed:
fixup! fixup! [AMDGPU] Allow casts between the Global and Constant Addr Spaces in isValidAddrSpaceCast

Disallow non-flat <-> (local, private) casts and use is_GlobalAddrSpace functions
